File: todolist-python/app/routes/file_routes.py
from flask import Blueprint, request, jsonify
import os
import time
from app.utils.cache_util import CacheUtil
from app.utils.config_util import config_util
from app.utils.logger_util import logger

bp = Blueprint('file_routes', __name__)
cache_util = CacheUtil()
DATA_DIR = config_util.get_data_dir()

FILE_OPTIONS = {
    "allow_expired": False,
    "ttl": 300000
}

def _scan_files():
    """扫描文件系统"""
    try:
        files = os.listdir(DATA_DIR)
        todo_files = [
            {
                "name": file,
                "exists": True,
                "mtime": int(os.stat(os.path.join(DATA_DIR, file)).st_mtime * 1000)
            }
            for file in files 
            if file.startswith('todo') and file.endswith('.txt')
        ]
        sorted_files = sorted(todo_files, key=lambda x: x.get("mtime", 0), reverse=True)
        return {
            "success": True, 
            "files": sorted_files, 
            "defaultFile": sorted_files[0]["name"] if sorted_files else "todo.txt"
        }
    except Exception as e:
        return {'success': False, 'error': str(e)}

def _read_file(filename):
    """读取文件内容"""
    file_path = os.path.join(DATA_DIR, filename)
    try:
        if not os.path.exists(file_path):
            with open(file_path, 'w', encoding='utf8') as f:
                f.write('')
        with open(file_path, 'r', encoding='utf8') as f:
            content = f.read()
        logger.info(f"读取文件 {filename}")
        return {"success": True, "content": content}
    except Exception as e:
        return {"success": False, "error": str(e)}

def _validate_filename(filename):
    """验证文件名"""
    return filename.startswith('todo') and filename.endswith('.txt')

def _clear_file_cache(specific_file=None):
    """清除文件缓存"""
    try:
        if specific_file:
            cache_util.delete(cache_util.generate_file_cache_key(specific_file))
        cache_util.delete('file_list')
    except Exception as e:
        logger.error(f'清除文件缓存失败: {str(e)}')

@bp.route('/scan', methods=['GET'])
def scan():
    try:
        cached_data = cache_util.get_wrapped_data('file_list', FILE_OPTIONS)
        if cached_data is not None:
            return jsonify(cached_data)
        
        data = _scan_files()
        response_data = {
            "success": data.get("success", True),
            "files": data.get("files", []),
            "defaultFile": data.get("defaultFile", "todo.txt")
        }
        cache_util.set_data('file_list', response_data, FILE_OPTIONS)
        return jsonify(cache_util.get_wrapped_data('file_list', FILE_OPTIONS))
    except Exception as e:
        return jsonify({"error": {"message": f"扫描文件失败: {str(e)}"}}), 500

@bp.route('/read/<filename>', methods=['GET'])
def read(filename):
    try:
        if not _validate_filename(filename):
            return jsonify({"success": False, "message": "不允许访问此文件"}), 403
        
        cache_key = cache_util.generate_file_cache_key(filename)
        cached_data = cache_util.get_wrapped_data(cache_key, FILE_OPTIONS)
        if cached_data is not None:
            return jsonify(cached_data)
        
        data = _read_file(filename)
        response_data = {
            "success": data.get("success", True),
            "content": data.get("content", "")
        }
        cache_util.set_data(cache_key, response_data, FILE_OPTIONS)
        return jsonify(cache_util.get_wrapped_data(cache_key, FILE_OPTIONS))
    except Exception as e:
        return jsonify({"error": {"message": f"读取文件失败: {str(e)}"}}), 500

@bp.route('/write/<filename>', methods=['POST'])
def write(filename):
    try:
        if not _validate_filename(filename):
            return jsonify({"success": False, "message": "不允许访问此文件"}), 403
        
        content = request.json.get('content', '')
        file_path = os.path.join(DATA_DIR, filename)
        
        with open(file_path, 'w', encoding='utf8') as f:
            f.write(content)
        
        _clear_file_cache(filename)
        logger.info(f"文件 {filename} 保存成功，相关缓存已清除")
        return jsonify({"success": True, "message": "文件保存成功"})
    except Exception as e:
        return jsonify({"success": False, "message": "写入文件失败", "error": str(e)}), 500

File: app/routes/file_routes.go
package routes

import (
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/gin-gonic/gin"
	"todolist-go/app/utils"
)

// FileOptions 缓存配置
var fileOptions = utils.CacheOptions{AllowExpired: false, TTL: 300000}

// RegisterFileRoutes 注册文件相关路由
func RegisterFileRoutes(router *gin.Engine) {
	fileGroup := router.Group("/api/file")
	fileGroup.GET("/scan", scanFiles)
	fileGroup.GET("/read/:filename", readFile)
	fileGroup.POST("/write/:filename", writeFile)
}

// scanFilesLogic 扫描文件系统
func scanFilesLogic() map[string]interface{} {
	dataDir := utils.ConfigUtilInstance.GetDataDir()
	files, err := os.ReadDir(dataDir)
	if err != nil {
		return map[string]interface{}{"success": false, "error": err.Error()}
	}

	var todoFiles []map[string]interface{}
	for _, file := range files {
		if !file.IsDir() && strings.HasPrefix(file.Name(), "todo") && strings.HasSuffix(file.Name(), ".txt") {
			fileData := map[string]interface{}{"name": file.Name(), "exists": true}
			if info, err := file.Info(); err == nil {
				fileData["mtime"] = info.ModTime().UnixMilli()
			}
			todoFiles = append(todoFiles, fileData)
		}
	}

	sort.Slice(todoFiles, func(i, j int) bool {
		mi, _ := todoFiles[i]["mtime"].(int64)
		mj, _ := todoFiles[j]["mtime"].(int64)
		return mi > mj
	})

	defaultFile := "todo.txt"
	if len(todoFiles) > 0 {
		defaultFile, _ = todoFiles[0]["name"].(string)
	}
	return map[string]interface{}{"success": true, "files": todoFiles, "defaultFile": defaultFile}
}

// scan 扫描文件接口
func scanFiles(c *gin.Context) {
	cachedData := utils.CacheUtilInstance.GetWrappedData("file_list", fileOptions)
	if cachedData != nil {
		c.JSON(http.StatusOK, cachedData)
		return
	}

	data := scanFilesLogic()
	responseData := map[string]interface{}{
		"success": data["success"], "files": data["files"], "defaultFile": data["defaultFile"],
	}
	utils.CacheUtilInstance.SetData("file_list", responseData, fileOptions)
	cachedData = utils.CacheUtilInstance.GetWrappedData("file_list", fileOptions)
	c.JSON(http.StatusOK, cachedData)
}

// readFileLogic 读取文件
func readFileLogic(filename string) map[string]interface{} {
	filePath := filepath.Join(utils.ConfigUtilInstance.GetDataDir(), filename)
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		os.WriteFile(filePath, []byte{}, 0644)
	}
	content, err := os.ReadFile(filePath)
	if err != nil {
		return map[string]interface{}{"success": false, "error": err.Error()}
	}
	return map[string]interface{}{"success": true, "content": string(content)}
}

// validateFilename 验证文件名
func validateFilename(filename string) bool {
	return strings.HasPrefix(filename, "todo") && strings.HasSuffix(filename, ".txt")
}

// clearFileCache 清除文件缓存
func clearFileCache(specificFile string) {
	if specificFile != "" {
		utils.CacheUtilInstance.Delete(utils.CacheUtilInstance.GenerateFileCacheKey(specificFile))
	}
	utils.CacheUtilInstance.Delete("file_list")
}

// read 读取文件内容接口
func readFile(c *gin.Context) {
	filename := c.Param("filename")
	if !validateFilename(filename) {
		c.JSON(http.StatusForbidden, map[string]interface{}{"success": false, "message": "不允许访问此文件"})
		return
	}

	cacheKey := utils.CacheUtilInstance.GenerateFileCacheKey(filename)
	cachedData := utils.CacheUtilInstance.GetWrappedData(cacheKey, fileOptions)
	if cachedData != nil {
		c.JSON(http.StatusOK, cachedData)
		return
	}

	data := readFileLogic(filename)
	responseData := map[string]interface{}{"success": data["success"], "content": data["content"]}
	utils.CacheUtilInstance.SetData(cacheKey, responseData, fileOptions)
	cachedData = utils.CacheUtilInstance.GetWrappedData(cacheKey, fileOptions)
	c.JSON(http.StatusOK, cachedData)
}

// write 写入文件内容接口
func writeFile(c *gin.Context) {
	filename := c.Param("filename")
	if !validateFilename(filename) {
		c.JSON(http.StatusForbidden, map[string]interface{}{"success": false, "message": "不允许访问此文件"})
		return
	}

	var req struct{ Content string `json:"content"` }
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, map[string]interface{}{"success": false, "message": "请求数据格式错误"})
		return
	}

	filePath := filepath.Join(utils.ConfigUtilInstance.GetDataDir(), filename)
	if err := os.WriteFile(filePath, []byte(req.Content), 0644); err != nil {
		c.JSON(http.StatusInternalServerError, map[string]interface{}{"success": false, "message": "写入文件失败", "error": err.Error()})
		return
	}

	clearFileCache(filename)
	c.JSON(http.StatusOK, map[string]interface{}{"success": true, "message": "文件保存成功"})
}
